### PR TITLE
Improve testing UX

### PR DESCRIPTION
**Pull request by xandkar**
TL;DR: ACS suite is slow because N is high.

NTL;R: Profiled the test suite and found that:
1. ACS suite takes the longest
2. in ACS suite two tests take the longest:
  - `init_test`
  - `one_dead_test`
3. in the above two tests, most time is spent in `do_send_outer`
4. overall most time is spent doing list appends
5. then, the main difference between ACS and other suites is the magnitude
   of default `N`, whereas ACS defaults to 34, others use a lower value
6. finally, decreasing ACS's N to <20 - dramatically speeds-up execution:

```sh
$ for n in $(seq 5 5 35); do printf "N=%d => " "$n"; time N="$n" ./rebar3 ct --suite hbbft_acs_SUITE > /dev/null; done
N=5 => N="$n" ./rebar3 ct --suite hbbft_acs_SUITE > /dev/null  3.38s user 0.51s system 107% cpu 3.611 total
N=10 => N="$n" ./rebar3 ct --suite hbbft_acs_SUITE > /dev/null  3.94s user 0.61s system 111% cpu 4.077 total
N=15 => N="$n" ./rebar3 ct --suite hbbft_acs_SUITE > /dev/null  5.14s user 0.72s system 109% cpu 5.335 total
N=20 => N="$n" ./rebar3 ct --suite hbbft_acs_SUITE > /dev/null  9.78s user 0.85s system 122% cpu 8.677 total
N=25 => N="$n" ./rebar3 ct --suite hbbft_acs_SUITE > /dev/null  22.26s user 1.30s system 149% cpu 15.761 total
N=30 => N="$n" ./rebar3 ct --suite hbbft_acs_SUITE > /dev/null  152.79s user 19.09s system 148% cpu 1:55.58 total
N=35 => N="$n" ./rebar3 ct --suite hbbft_acs_SUITE > /dev/null  514.02s user 57.72s system 134% cpu 7:06.10 total
```

In the long-run we should probably revisit implementation of `do_send_outer`.

### CODE REVIEW

**Comment by xandkar** [comment]
I removed all the manual profiling artifacts, but left the `eprof` stuff in ACS suite, because it is still asymptotically slow and it would likely be useful later on when someone runs it with a high N again and be surprised by the running time.

**Comment by Vagabond** [comment]
Yeah, do_send_outer is slow. I think it's ok to be slow but correct in this case though. In general we don't need high N for testing as it doesn't really find any more bugs than testing with N=4.

This all seems fine.

File: test/hbbft_test_utils.erl [hunk 1]
```diff
@@ -2,6 +2,13 @@
 
 -export([do_send_outer/4, shuffle/1, random_n/2, enumerate/1, merge_replies/3]).
 
+% TODO Type of Acc elements
+% TODO Type of States elements
+% TODO Type of Results elements
+-spec do_send_outer(Module :: atom(), Results :: list(), States, Acc) ->
```
Comment (Vagabond) [inline test/hbbft_test_utils.erl:8]: Specing a test-only function is kind of useless without some dialyzer contortions to run the dialyzer on test code (which can end up with a lot of noise because tests often intentionally call code with bad arguments).

**Review by Vagabond — APPROVED** [review]

File: test/hbbft_test_utils.erl (hunk 1 above)
Comment (xandkar) [reply test/hbbft_test_utils.erl:8]: @Vagabond Make sense. Do you know how to do that through `rebar`? I see no obvious way other than adding a manual `dialyzer` call to the make recipe, like `dialyzer test/hbbft_test_utils.erl`

Comment (Vagabond) [reply test/hbbft_test_utils.erl:8]: You can configure the dialyzer to scan test code, somehow, but I don't think it is worth it.

Comment (xandkar) [reply test/hbbft_test_utils.erl:8]: Gotcha. I was mainly thinking along the lines of documentation, since the spec is not immediately obvious. Alternatively, we could add it as a doc comment, but a spec is something that can at least potentially be checked, so a superior option, IMO.

Comment (xandkar) [reply test/hbbft_test_utils.erl:8]: @Vagabond Added dialyzer to test recipe. Thanks, @evanmcc!
